cpp/src/strings/replace/multi.cu

```diff
@@ -0,0 +1,488 @@
+
+  // create a vector of offsets to each string's set of target positions
+  auto const targets_offsets = [&] {
+    auto string_indices = rmm::device_uvector<size_type>(target_count, stream);
+
+    auto const pos_itr = cudf::detail::make_counting_transform_iterator(
+      0, [d_positions] __device__(auto idx) -> size_type { return d_positions[idx].first; });
+    auto pos_count = std::distance(d_positions, copy_end);
+
+    thrust::upper_bound(rmm::exec_policy(stream),
+                        input.offsets_begin(),
+                        input.offsets_end(),
+                        pos_itr,
+                        pos_itr + pos_count,
+                        string_indices.begin());
+
+    // compute offsets per string
+    auto targets_offsets   = rmm::device_uvector<size_type>(strings_count + 1, stream);
+    auto d_targets_offsets = targets_offsets.data();
+
+    // memset to zero-out the target counts for any null-entries or strings with no targets
+    CUDF_CUDA_TRY(cudaMemsetAsync(
+      d_targets_offsets, 0, targets_offsets.size() * sizeof(size_type), stream.value()));
+
+    // next, count the number of targets per string
+    auto d_string_indices = string_indices.data();
+    thrust::for_each_n(rmm::exec_policy(stream),
+                       thrust::make_counting_iterator<size_type>(0),
+                       target_count,
+                       [d_string_indices, d_targets_offsets] __device__(size_type idx) {
+                         auto const str_idx = d_string_indices[idx] - 1;
+                         atomicAdd(d_targets_offsets + str_idx, 1);
+                       });
+    // finally, convert the counts into offsets
+    thrust::exclusive_scan(rmm::exec_policy(stream),
+                           targets_offsets.begin(),
+                           targets_offsets.end(),
+                           targets_offsets.begin());
+    return targets_offsets;
+  }();
+  auto const d_targets_offsets = targets_offsets.data();
+
+  // compute the number of string segments produced by replace in each string
+  auto counts = rmm::device_uvector<size_type>(strings_count, stream);
+  thrust::transform(rmm::exec_policy(stream),
+                    thrust::make_counting_iterator<size_type>(0),
+                    thrust::make_counting_iterator<size_type>(strings_count),
+                    counts.begin(),
+                    [fn, d_positions, d_targets_offsets] __device__(size_type idx) -> size_type {
+                      return fn.count_strings(idx, d_positions, d_targets_offsets);
+                    });
+
+  // create offsets from the counts
+  auto offsets =
+    std::get<0>(cudf::detail::make_offsets_child_column(counts.begin(), counts.end(), stream, mr));
+  auto const total_strings =
+    cudf::detail::get_value<size_type>(offsets->view(), strings_count, stream);
+  auto const d_strings_offsets = offsets->view().data<size_type>();
```

Comment:
```suggestion
  auto const d_strings_offsets = offsets->view().data<offset_type>();
```

Reply:
Trying to move away from `offset_type` in favor of using `size_type` everywhere.

Reply:
I hadn't heard that yet. I guess that means we aren't considering 64-bit offsets any time soon.

Reply:
@hyperbolic2346 It's actually a slightly different concern. Currently `offset_type` is defined to be the same as `size_type`. If the two were defined differently, however, it would cause a lot of issues. The library design and API return types pretty much require that `offset_type == size_type` regardless of whether it's a 32 or 64 bit value. We should be moving away from `offset_type` in favor of `size_type` everywhere. (Perhaps we should do a mass find-replace?)

Reply:
I know that there would be large issues if they two were ever not equal, but I thought the design intent was that they *could* be different sizes, and if we decided to do that we would require a good deal of cleanup to remove the cross-contamination.

I was under the, seemingly false, impression that the goal was to differentiate the two so we could switch the types.